rpcs3/Emu/Cell/lv2/sys_game.cpp

```diff
@@ -5,52 +5,106 @@
 #include "Emu/System.h"
 #include "Emu/system_config.h"
 
-#include <thread>
+#include "Emu/IdManager.h"
+#include "Utilities/Thread.h"
 
 #include "sys_game.h"
 
 LOG_CHANNEL(sys_game);
 
-atomic_t<bool> watchdog_stopped = true;
-atomic_t<u64> watchdog_last_clear;
-u64 get_timestamp()
+struct watchdog_t
 {
-	return (get_system_time() - Emu.GetPauseTime());
-}
+	struct alignas(8) control_t
+	{
+		bool needs_restart = false;
+		bool active = false;
+		char pad[sizeof(u32) - sizeof(bool) * 2]{};
+		u32 timeout = 0;
+	};
 
-error_code _sys_game_watchdog_start(u32 timeout)
-{
-	sys_game.trace("sys_game_watchdog_start(timeout=%d)", timeout);
+	atomic_t<control_t> control;
 
-	if (!watchdog_stopped)
+	void operator()()
 	{
-		return CELL_EABORT;
-	}
+		u64 start_time = get_system_time();
+		u64 old_time = start_time;
+		u64 current_time = old_time;
 
-	auto watchdog = [=]()
-	{
-		while (!watchdog_stopped)
+		constexpr u64 sleep_time = 50'000;
+
+		while (thread_ctrl::state() != thread_state::aborting)
 		{
-			if (Emu.IsStopped() || get_timestamp() - watchdog_last_clear > timeout * 1000000)
+			if (Emu.IsPaused())
```

Comment:
The watchdog wasn't triggered when the game status was "frozen" because Emu.IsPaused() includes not only `paused`, but also `frozen`, `ready`, and `starting`. Shouldn't the watchdog attempt to restart the game when the game is no longer responding?

Reply:
Hmm, I need to think what to do.

Reply:
Fixed.

Reply:
Great. It was triggered in both cases.